torch/csrc/jit/codegen/cuda/fusion.cpp

```diff
@@ -536,22 +536,41 @@ bool Fusion::hasReduction() {
 std::vector<Val*> Fusion::getTerminatingOutputs() {
   FUSER_PERF_SCOPE("getTerminatingOutputs");
 
-  std::unordered_set<Val*> used_vals;
-
-  const auto exprs = ExprSort::getExprs(
-      this, std::vector<Val*>(outputs().begin(), outputs().end()));
-
-  for (auto expr : exprs) {
-    for (auto inp : expr->inputs())
-      used_vals.emplace(inp);
-  }
+  auto is_reachable_to_output = [](Val* val) {
+    // traverse to consumers of val and see if there is an output
+    std::deque<Val*> consumers;
+    for (auto use : val->uses()) {
+      for (auto consumer : use->outputs()) {
+        consumers.push_back(consumer);
+      }
+    }
+    while (!consumers.empty()) {
+      auto consumer = consumers.front();
+      consumers.pop_front();
+      if (consumer->isFusionOutput()) {
+        return true;
+      }
+      // consumer is not an output; proceed to its consumers
+      for (auto use : consumer->uses()) {
+        for (auto consumer_of_consumer : use->outputs()) {
+          consumers.push_back(consumer_of_consumer);
```

Comment:
If this was set to `push_front` instead of `push_back` would this be DFS then and potentially a lot faster at finding an output? (Would be the same thing if above was set to `consumers.back()` and `consumers.pop_back()`

Reply:
I don't think it would be a lot different anyway as the number of use expressions would be small, but yes I'd expect DFS would be faster. I'll change it to do `consumers.back()`. Thanks fort the feedback.